src/did.rs

```diff
@@ -384,6 +385,102 @@ pub struct DIDParameters {
     pub property_set: Option<Map<String, Value>>,
 }
 
+/// DID Create Operation
+///
+/// <https://identity.foundation/did-registration/#create>
+pub struct DIDCreate {
+    pub update_key: Option<JWK>,
+    pub recovery_key: Option<JWK>,
+    pub verification_key: Option<JWK>,
+    pub options: Value,
+}
+
+/// DID Update Operation
+///
+/// <https://identity.foundation/did-registration/#update>
+pub struct DIDUpdate {
+    pub did: String,
+    pub update_key: Option<JWK>,
+    pub new_update_key: Option<JWK>,
+    pub operation: DIDDocumentOperation,
+    pub options: Value,
+}
+
+/// DID Recover Operation
+///
+/// <https://www.w3.org/TR/did-core/#did-recovery>
+pub struct DIDRecover {
+    pub did: String,
+    pub recovery_key: Option<JWK>,
+    pub new_update_key: Option<JWK>,
+    pub new_recovery_key: Option<JWK>,
+    pub new_verification_key: Option<JWK>,
+    pub options: Value,
+}
+
+/// DID Deactivate Operation
+///
+/// <https://identity.foundation/did-registration/#deactivate>
+pub struct DIDDeactivate {
+    pub did: String,
+    pub key: Option<JWK>,
+    pub options: Value,
+}
+
+/// DID Document Operation
+///
+/// This should represent [didDocument][dd] and [didDocumentOperation][ddo] specified by DID
+/// Registration.
+///
+/// [dd]: https://identity.foundation/did-registration/#diddocumentoperation
+/// [ddo]: https://identity.foundation/did-registration/#diddocument
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(tag = "didDocumentOperation", content = "didDocument")]
+#[serde(rename_all = "camelCase")]
+pub enum DIDDocumentOperation {
+    /// Set the contents of the DID document
+    ///
+    /// setDidDocument operation defined by DIF DID Registration
+    SetDidDocument(Document),
+
+    /// Add properties to the DID document
+    ///
+    /// addToDidDocument operation defined by DIF DID Registration
+    AddToDidDocument(HashMap<String, Value>),
+
+    /// Remove properties from the DID document
+    ///
+    /// removeFromDidDocument operation defined by DIF Registration
+    RemoveFromDidDocument(Vec<String>),
+
+    /// Add or update a verification method in the DID document
+    SetVerificationMethod {
+        vmm: VerificationMethodMap,
+        purposes: Vec<VerificationRelationship>,
+    },
+
+    /// Add or update a service map in the DID document
+    SetService(Service),
+
+    /// Remove a verification method in the DID document
+    RemoveVerificationMethod(DIDURL),
+
+    /// Add or update a service map in the DID document
+    RemoveService(DIDURL),
+}
+
+/// A transaction for a DID method
+#[derive(Debug, Serialize, Deserialize, Builder, Clone, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct DIDMethodTransaction {
```

Comment:
More info about the proposed transaction concept, in didkit-cli's readme: https://github.com/spruceid/didkit/blob/3a220b0fddfa6565a2214284430a079ac7057a10/cli/README.md#did-method-transaction
DID Method Transaction is like a [Sidetree Operation Request](https://identity.foundation/sidetree/api/#sidetree-operations) generalized to arbitrary DID methods. The contents are DID-method-specific, except for the `didMethod` property which would be used by DIDKit and/or ssi's DIDMethods struct to route the operation to a DID method implementation for submission/processing.
In terms of DID Registration, the transaction represents a DID operation that is ready to be written to the DID method's verifiable data registry. I don't see this data structure as being exposed in DID Registration API; it's otherwise internal to the 
implementation. It's not using DID Registration's [didDocument](https://identity.foundation/did-registration/#diddocument)/[didDocumentOperation](https://identity.foundation/did-registration/#diddocumentoperation) fields - although it might be convertible to/from them - but its signature is over a method-specific data structure. For other DID methods, the DID method transaction might contain a signed blockchain transaction (e.g. for `did:ethr`, `did:tz`, and `did:btcr`). (For Sidetree, the blockchain transaction is a separate step, in this implementation assumed to be done on the other side of the Sidetree REST API.)

Sidetree in https://identity.foundation/sidetree/spec/v1.0.0/#sidetree-operations requires clients to "keep the operation payload once it is submitted to a Sidetree node until it is generally available and observed" and to resubmit the same operation payload "if the submitted operation is not anchored and propagated, for whatever reason". Without the transaction concept, the operation functions here would need to block until the operation is observed, and/or maintain the operation in some external state. By returning the Sidetree Operation request as a DID method transaction to be separately submitted/published using `DIDMethod::submit_transaction` (`didkit did-submit-tx`), the operation functions stay more side-effect-free and do not risk being interrupted and losing the operation. Keeping the signed operations also allows the caller to prove their DID state with the chain of signed operations.

Note that currently this implementation does not wait for and detect that an operation has been successfully applied. So the caller must do that, resubmitting the operation if needed. This limitation could be addressed by making the submit tx function poll the resolver to detect the corresponding update - or by adding a separate function to do that.

I might need help to make the transaction into a trait. I thought of making the transaction type into an associated type on DIDMethod, but this does not work, because DIDResolver is used as a trait object, so DIDMethod cannot have generic type parameters, e.g.:
```
error[E0038]: the trait `DIDResolver` cannot be made into an object
    --> src/ldp.rs:2085:20
     |
2085 |         resolver: &dyn DIDResolver,
     |                    ^^^^^^^^^^^^^^^ `DIDResolver` cannot be made into an object
     |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visi
t <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
    --> src/did_resolve.rs:469:8
     |
410  | pub trait DIDResolver: Sync {
     |           ----------- this trait cannot be made into an object...
...
469  |     fn to_did_method<T>(&self) -> Option<&dyn DIDMethod<Transaction = T>> {
     |        ^^^^^^^^^^^^^ ...because method `to_did_method` has generic type parameters
     = help: consider moving `to_did_method` to another trait
```
Also I'm not sure how this would be handled in DIDMethods where there is a map of strings to DIDMethod trait objects which each might use a different transaction type, but I'm not sure how to express that:
```
pub struct DIDMethods<'a> {
    /// Collection of DID methods by method id.
    pub methods: HashMap<&'a str, &'a dyn DIDMethod<Transaction = ...>>,
}
```
If the functions using the Transaction type are moved out of DIDMethod into a new trait, e.g. DIDRegistrar or DIDOperator (like how the resolve functions are in DIDResolver), I would still assume to use a conversion function like `fn to_registrar<T>(&self) -> &dyn DIDRegistrar<Transaction = T>` (like `DIDMethod::to_resolver`) but this still doesn't work with DIDMethod being object-safe, so again I'm not sure what to do.

Reply:
TBH I think `to_resolver` is an anti-pattern. If you need both `DIDMethod` and `DIDResolver` they should be both specified or if you can't have one without the other they should be combined.

I like the idea of having multiple implementations of a `DIDRegistar` trait for each method for each kind of transaction. It puts the burden on the user of `ssi` to route the types of messages, but it probably make sense? Some DIDs can be created generated locally and/or created in a registrar -- there are so many variants that users are probably best placed to chose what to do?